### PR TITLE
feature(Datepicker): keyboard input of date

### PR DESCRIPTION
**Pull request by cwklausing**
Allows calendar to match current AG functionality. Backwards compatible--just adds extra (optional) prop.

Current AG datepicker requires moment, which we're removing, so we're updating this one to match AG needs. [Card](https://rentpath.leankit.com/card/697016601)

### CODE REVIEW

File: src/Datepicker.js [hunk 1]
```diff
@@ -68,6 +73,24 @@ export default class Datepicker extends Component {
     })
   }
 
+  handleChange(e) {
```
Comment (tadjohnston) [inline src/Datepicker.js]: Since you've change this to a controlled component, what happens if I pass in a `value` prop for the input after the component has been mounted?

File: src/Datepicker.js [hunk 2]
```diff
@@ -18,10 +20,12 @@ export default class Datepicker extends Component {
     this.handleCalendarVisibility = this.handleCalendarVisibility.bind(this)
     this.dateChange = this.dateChange.bind(this)
     this.outsideHandler = null
+    this.handleChange = this.handleChange.bind(this)
 
     this.state = {
       showCalendar: false,
       selectedDate: this.props.startDate,
+      textInput: '',
```
Comment (tadjohnston) [inline src/Datepicker.js]: Shouldn't this take in the default `value` prop instead?

File: src/Datepicker.js (hunk 1 above)
Comment (cwklausing) [reply src/Datepicker.js]: You wouldn't be able to pass down a new startDate (I assume that's what you mean by a `value` prop) once the component mounts. However, as best I can tell, this is consistent with the current date-picker functionality.

File: src/Datepicker.js (hunk 2 above)
Comment (cwklausing) [reply src/Datepicker.js]: Again, leaving it blank initially is consistent with current Prod datepicker functionality. But I agree that what you suggests makes more sense.

File: __tests__/Datepicker.js [hunk 3]
```diff
@@ -1,26 +1,74 @@
...
+
+test('if passed enableTextInput, should respect mindate & maxdate', () => {
+  const startDate = new Date('2020-01-04T06:00:00.000Z')
+  const datepicker = mount(<Datepicker dateFormat="MM/DD/YYYY" startDate={startDate} minDate={startDate} maxDate={addDays(startDate, 10)} enableTextInput/>)
+  const input = datepicker.find('input')
+  
+  input.simulate('change', {target: {value: '01/10/2020'}});
+  expect(datepicker.state().selectedDate).toBe('01/10/2020')
+  
+  input.simulate('change', {target: {value: '01/20/2020'}});
```
Comment (lfender6445) [inline __tests__/Datepicker.js]: changing to `01/19/2020` would work though right?

File: src/Datepicker.js [hunk 4]
```diff
@@ -68,6 +73,24 @@ export default class Datepicker extends Component {
...
 
+  handleChange(e) {
+    const text = e.target.value
+    this.setState({
+      textInput: text
+    })
+    const dateRegex = /\d{2}\/\d{2}\/\d{4}/
+
+    // check length first to avoid doing excessive regex checks
+    if(text.length === 10 && dateRegex.test(text)) {
```
Comment (lfender6445) [inline src/Datepicker.js]: is there not something like this built into `date-fns`

Comment (cwklausing) [reply src/Datepicker.js]: As best I can tell, no--there's nothing to check the actual _format_ of the date string, just the contents.

File: __tests__/Datepicker.js (hunk 3 above)
Comment (cwklausing) [reply __tests__/Datepicker.js]: No, the startDate is actually 1/4/2020, so 1/14 and before would work.

Comment (mpelzsherman) [reply __tests__/Datepicker.js]: It's generally best practice to test the boundary conditions, so I'd suggest using 1/15 here.

File: src/Datepicker.js [hunk 5]
```diff
@@ -68,6 +73,24 @@ export default class Datepicker extends Component {
...
+      const theDate = new Date(text)
+      const isDisabled = !isValid(theDate) ||
+        isBefore(theDate, new Date(this.props.minDate)) ||
+        isAfter(theDate, new Date(this.props.maxDate))
+      
+      if(!isDisabled) { this.dateChange(text) }
```
Comment (mpelzsherman) [inline src/Datepicker.js]: Shouldn't there be a space after `if` here, and on line 84?

**Comment by lfender6445** [comment]
👍
